### PR TITLE
AK: Introduce IMath.h for integral algorithms, starting with pow<I> and use it where applicable

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
This was mainly done to resolve:
```c++
#ifdef KERNEL
        // We don't have pow() in kernel land
        u64 scale = 10;
        for (size_t i = 0; i < precision - 1; i++) // TODO: not efficient
            scale *= 10;
#else
        u64 scale = pow(10.0, (double)precision);
#endif
```
But turns out we were doing the same in other parts of the code base aswell

### CODE REVIEW

**Comment by ADKaster** [comment]
Small nitpick but when I read "IMath.h" I expect it will contain some abstract base class called IMath with some pure virtuals for math functions. Maybe I spend too much time looking at pure OO codebases. Maybe "IntegralMath.h" would get its purpose across clearer?

File: AK/IntegerMath.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    if (exponent < 0)
+        return 0;
+
+    I res = 0;
+    while (exponent) {
+        if (exponent & 1)
+            res *= base;
+        res *= res;
+        exponent >>= 1;
+    }
```
Comment (BertalanD) [inline AK/IntegerMath.h:39]: This doesn't look right. I'm pretty sure this function will return 0 regardless of the inputs.

```suggestion
    I res = 1;
    while (exponent) {
        if (exponent & 1)
            res *= base;
        base *= base;
        exponent >>= 1;
    }
```

**Review by BertalanD — CHANGES_REQUESTED** [review]

File: AK/IntegerMath.h (hunk 1 above)
Comment (Hendiadyoin1) [reply AK/IntegerMath.h]: Yep....
Took me a while to figure the base * base out....

should work now, I hope

File: AK/IntegralMath.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+template<Integral T>
+constexpr T log2(T x)
+{
+    return x ? (8 * sizeof(T) - 1) - count_leading_zeroes(static_cast<MakeUnsigned<T>>(x)) : 0;
+}
+
+template<Integral I>
+inline I pow(I base, I exponent)
```
Comment (BertalanD) [inline AK/IntegralMath.h:30]:
```suggestion
constexpr I pow(I base, I exponent)
```

File: AK/IntegralMath.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        return 1;
+
+    I res = 1;
+    while (exponent > 1) {
+        if (exponent & 1)
+            res *= base;
+        base *= base;
+        exponent /= 2u;
+    }
+    return res * base;
```
Comment (BertalanD) [inline AK/IntegralMath.h:45]: We can avoid the extra `exponent == 0` branch if we tweak the loop condition a bit:

```suggestion
    if (exponent < 0)
        return 0;

    I res = 1;
    while (exponent > 0) {
        if (exponent & 1)
            res *= base;
        base *= base;
        exponent /= 2u;
    }
    return res;
```

**Review by BertalanD — CHANGES_REQUESTED** [review]
Small nitpick, otherwise LGTM!

**Comment by Panky-codes** [comment]
​There  is `round_up_to_power_of_two​` and `is_power_of_2` that operates on a integer in StdLibExtras.h. I think they also belong to this IntergalMath.h. Let me know your thoughts.

**Comment by Hendiadyoin1** [comment]
> ​There is `round_up_to_power_of_two​` and `is_power_of_2` that operates on a integer in StdLibExtras.h. I think they also belong to this IntergalMath.h. Let me know your thoughts.

This would incur further include changes
Feel free to make a PR for it

**Review by BertalanD — APPROVED** [review]

**Pull request merged**